### DISCUSSION TITLE
Interface vs Types

### DISCUSSION THREAD

**Discussion by yukinoda** [monstar-lab-oss/ml-frontend#20] [General]
This is the conclusion we had in the admin panel template discussion
https://github.com/monstar-lab-oss/admin-panel-template-reactjs/discussions/88#discussioncomment-942054

**Comment by jinmayamashita** [comment]
my opinion is Props:

Reason, discriminated union cannot be expressed using an interface.
so as an example:

```ts
// Props is a discriminated union of 2 object type
type Props = {
    isAdmin?: false, adminData?: never
} | {
    isAdmin: true, adminData: {
        foo: string,
        bar: number
    }
}

//  TYPE ERROR: because when isAdmin is false, it does not have adminData attribute.
const props: Props  = {
    isAdmin: false,
    adminData: {
        foo: 'foo',
        bar: 123
    }
}
```

[TypeScript Playground](https://www.typescriptlang.org/play?#code/PTAEAUCcHsAcGdQEtEENQBMUGNJILZIB2qALgKYagCuRS0Ro0AZqAExMBGAVudqaFIBPWOQBQw0RBgJQAXlABvMaFXJ4AQQyEiAfgBcoZqgA28cgBpQqbcQAiZVAdBFyAN3KQxAX1AAfJRU1FC0dQ1JIaktrWyIHUlRDZTUUo2hoQ3gI4gBzCyDU0E5USEMianxOTwLQbx8xMRBVABUATXAAUVAOgCUegHkewyrsVGpzUAB3AAtyRhDY9SNTcyskAQxockQiaAFp1A8YnXj0MmzOagoAOjFsBizQWBl4Qyg4RHlAlIWw5bNLDUbCdHEkailmOlDABySHQaH5QqqYqlUAARjYAGYanU6o0wMQKJBjNhyKB8KghNYzNAimTYNB4PAkJwTGTJutpoJZmkTCZoByiDkpnhSLlQFkhGzdKBrnLQIA+DcAKLugfHIIhEklkgCS7wQ2s0sWakTJyiaP0NOmcxgBVmB9kczlcHi8TTqTUJni1oF1LwNoWIADEVqa1RaA0Rwia7bFTmDzUi4ZlskL8gnCiiyhUqq6wCl3WAC6B7kRHs8PmxDL6Pv6jSb-D69fBazpgwDVAozXngpbiIYbasw2p7XFQYF06kk6BYekEUOMyVDBjsROi94gA)

```ts
// interface may also be possible with the following writing style? ... 🤔 
interface IPropsIsAdminTrue {
    isAdmin?: false, adminData?: never
}
interface IPropsIsAdminFalse {
    isAdmin: true, adminData: {
        foo: string,
        bar: number
    }
}
const props: IPropsIsAdminTrue | IPropsIsAdminFalse
```
